File: framework/languages/java/e2e/RegistrationCodec/Client/src/main/java/systems/zlink/e2e/registrationcodec/client/Scenarios/AutoRegistrationScenario.java
package systems.zlink.e2e.registrationcodec.client.Scenarios;

import java.time.Duration;
import systems.zlink.e2e.registrationcodec.client.Support.ScenarioAssert;
import systems.zlink.e2e.registrationcodec.client.Support.ScenarioContext;
import systems.zlink.e2e.registrationcodec.shared.Contracts;

public final class AutoRegistrationScenario {
    private static final Duration REQUEST_TIMEOUT = Duration.ofSeconds(5);

    private AutoRegistrationScenario() {
    }

    public static void run(ScenarioContext context) {
        Contracts.EchoRes auto = context.client().requestToChannel(
                Contracts.CHANNEL,
                new Contracts.EchoAutoReq("auto-request"))
            .packetName("EchoAuto")
            .timeout(REQUEST_TIMEOUT)
            .await(Contracts.EchoRes.class);
        ScenarioAssert.ensure("echo:auto-request".equals(auto.value()) && "auto".equals(auto.handler()),
            "RC-A1 request mismatch");
        context.client().sendToChannel(Contracts.CHANNEL, new Contracts.EchoAutoMsg("auto-send"))
            .packetName("EchoAuto")
            .await();
        ScenarioAssert.waitForEvidence(context.evidence(), "Send", "EchoAuto", "auto-send");
        System.out.println("scenario RC-A1 passed");
    }
}

File: framework/languages/java/e2e-kotlin/RegistrationCodec/Client/src/main/kotlin/systems/zlink/e2e/kotlin/registrationcodec/client/Scenarios/AutoRegistrationScenario.kt
package systems.zlink.e2e.kotlin.registrationcodec.client.scenarios

import systems.zlink.e2e.kotlin.registrationcodec.Contracts
import systems.zlink.e2e.kotlin.registrationcodec.EchoAutoMsg
import systems.zlink.e2e.kotlin.registrationcodec.EchoAutoReq
import systems.zlink.e2e.kotlin.registrationcodec.EchoAutoRes
import systems.zlink.e2e.kotlin.registrationcodec.client.support.EvidenceText
import systems.zlink.e2e.kotlin.registrationcodec.client.support.ScenarioAssert
import systems.zlink.framework.channels.ZLinkClient

class AutoRegistrationScenario(
    private val client: ZLinkClient,
    private val evidence: EvidenceText,
    private val assert: ScenarioAssert,
) {
    fun run() {
        val auto = client.requestToChannel(Contracts.CHANNEL, EchoAutoReq("auto-request"))
            .packetName("EchoAutoReq")
            .timeout(requestTimeout)
            .await(EchoAutoRes::class.java)
        assert.that(auto.value == "echo:auto-request" && auto.handler == "auto", "RC-A1 request mismatch")
        client.sendToChannel(Contracts.CHANNEL, EchoAutoMsg("auto-send"))
            .packetName("EchoAutoMsg")
            .await()
        evidence.waitForEvidence("Send", "EchoAutoMsg", "auto-send")
        println("scenario RC-A1 passed")
    }
}
